transports/webrtc/src/connection.rs

```diff
@@ -0,0 +1,280 @@
+            }))
+            .await;
+    }
+}
+
+impl<'a> StreamMuxer for Connection {
+    type Substream = PollDataChannel;
+    type Error = Error;
+
+    fn poll_inbound(&self, cx: &mut Context<'_>) -> Poll<Result<Self::Substream, Self::Error>> {
+        let mut inner = self.inner.lock().unwrap();
+        match ready!(inner.incoming_data_channels_rx.poll_next(cx)) {
+            Some(detached) => {
+                trace!("Incoming substream {}", detached.stream_identifier());
+
+                let mut ch = PollDataChannel::new(detached);
+                if let Some(cap) = inner.read_buf_cap {
+                    ch.set_read_buf_capacity(cap);
+                }
+
+                Poll::Ready(Ok(ch))
+            }
+            None => Poll::Ready(Err(Error::InternalError(
+                "incoming_data_channels_rx is closed (no messages left)".to_string(),
+            ))),
+        }
+    }
+
+    fn poll_address_change(&self, _cx: &mut Context<'_>) -> Poll<Result<Multiaddr, Self::Error>> {
+        return Poll::Pending;
+    }
+
+    fn poll_outbound(&self, cx: &mut Context<'_>) -> Poll<Result<Self::Substream, Self::Error>> {
+        let mut inner = self.inner.lock().unwrap();
+        let peer_conn = self.peer_conn.clone();
+        let fut = inner.outbound_fut.get_or_insert(Box::pin(async move {
+            let peer_conn = peer_conn.lock().await;
+
+            // Create a datachannel with label 'data'
+            let data_channel = peer_conn
+                .create_data_channel("data", None)
+                .map_err(Error::WebRTC)
+                .await?;
+
+            trace!("Opening outbound substream {}", data_channel.id());
+
+            // No need to hold the lock during the DTLS handshake.
+            drop(peer_conn);
+
+            let (tx, rx) = oneshot::channel::<Arc<DetachedDataChannel>>();
+
+            // Wait until the data channel is opened and detach it.
+            register_data_channel_open_handler(data_channel, tx).await;
+
+            // Wait until data channel is opened and ready to use
+            match rx.await {
+                Ok(detached) => Ok(detached),
+                Err(e) => Err(Error::InternalError(e.to_string())),
+            }
+        }));
+
+        match ready!(fut.as_mut().poll(cx)) {
+            Ok(detached) => {
+                let mut ch = PollDataChannel::new(detached);
+                if let Some(cap) = inner.read_buf_cap {
+                    ch.set_read_buf_capacity(cap);
+                }
+
+                Poll::Ready(Ok(ch))
+            }
+            Err(e) => Poll::Ready(Err(e)),
+        }
+    }
+
+    fn poll_close(&self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        debug!("Closing connection");
+
+        let mut inner = self.inner.lock().unwrap();
+        let peer_conn = self.peer_conn.clone();
+        let fut = inner.close_fut.get_or_insert(Box::pin(async move {
+            let peer_conn = peer_conn.lock().await;
```

Comment:
Is the following assumption correct?

The `Future` in `inner.outbound_fut` might be holding this lock. Given that it is a `futures::lock::Mutex`, the `inner.outbound_fut` might hold the lock across `.await` points, in other words potentially for a longer period of time. In this particular case as long as `peer_conn.create_data_channel.await` in `inner.outbound_fut` takes.

In this particular case, given that we want to close the connection anyways, we are no longer interested in the result of `inner.outbound_fut`. Thus I think it is fine to drop `inner.outbound_fut` and thus unlocking the lock.

In general, I find reasoning about concurrent code using locks to add a lot of mental complexity. E.g. in this case I need to understand what `peer_conn.create_data_channel` does, and what it in turn depends on, in order to be sure whether it is ok to hold a lock across `peer_conn.create_data_channel.await`.

Does that reasoning make sense @melekes? Do you agree that this adds complexity? It might be the case that this is only hard for me, though not hard for others.

Reply:
> Does that reasoning make sense @melekes? Do you agree that the usage of asynchronous locks adds a large amount of complexity here? It might be the case that this is only hard for me, though not hard for others.

Yep. I agree that locks add complexity. The alternative is to rewrite Connection similarly to https://github.com/libp2p/rust-libp2p/pull/2622#discussion_r935377251 to use flume channels and handles. The amount of complexity seems to be the same to me in both cases, while 1 lock shared between two futures results in more concise code.